### DISCUSSION TITLE
reporting BodyStream errors?

### DISCUSSION THREAD

**Discussion by bburdette** [actix/actix-web#3265] [Q&A] [answered]
I'm passing a `Stream<Result<Bytes, Box<std::error::Error>>>` into `HttpResponse::Ok().streaming(mystream)`.  It works, but if there's an error the stream simply stops, and nothing gets reported on either the client or the server side.  

I could use `map` on the stream to transform errors into Bytes format messages.  That would give me error reporting on the client side.

But probably what I want is to log the error on the server side.  It's not immediately obvious how to do this.  Suggestions?

**Comment by robjtede** [comment]
Errors resulting from response body streams yielding `Err(_)` are logged in both HTTP/1.x and HTTP/2 code-paths:

- https://github.com/actix/actix-web/blob/master/actix-http/src/h1/dispatcher.rs#L1190
- https://github.com/actix/actix-web/blob/master/actix-http/src/h2/dispatcher.rs#L147-L157

These logs can be viewed with the appropriate `log` logger such as `env_logger`.

The main issue is that there's nothing really sensible to do with streaming errors other than log the error, close the connection/stream, and hope the client retries. Printing the error to the stream before closing it is not common and not something we'd want to do in general.

**Reply by bburdette** [reply]
Unfortunately I don't see those on my server.  I have a function that I use to transform my stream from my SyncMessage struct to Bytes:

```
pub fn bytesify(
  x: Result<SyncMessage, Box<dyn std::error::Error>>,
) -> Result<Bytes, Box<dyn std::error::Error>> {
  error!("log an error message here!");

  // instead of converting to bytes, fail for testing
  Err(zkerr::Error::String("this is an error in the stream".into()).into())
}
```

I apply that to my stream with map.  This version of it fails right away.  In stdio I see the "log an error message here!" message, but I don't see the "this is an error in the stream".  The client gets disconnected as one would expect.

What's reported in the last couple lines of stdio:

```
 INFO 2024-01-26T21:02:08Z: tracing_actix_web::root_span_builder: HTTP request; http.method=POST http.route=/stream http.flavor=1.1 http.scheme=http http.host=localhost:8010 http.client_ip=127.0.0.1 http.user_agent= http.target=/stream otel.name=HTTP POST /stream otel.kind="server" request_id=b49bf899-aae2-493c-915e-fed245383915
priavte_streaming item: Json(PrivateStreamingMessage { what: Sync, data: Some(Object {"after": Number(1706299917000)}) })
ERROR 2024-01-26T21:02:08Z: zknotes_server_lib::sync: log an error message here!
```

**Reply by bburdette** [reply]
I guess I've answered my own question about how to log these errors, since I apply bytesify to all my streams I can just log the error there.  

But sounds like actix is supposed to do that for me, maybe its a bug?  I'm on actix-web 4.3.1 fwiw.

**Reply by robjtede** [reply, marked as answer]
I checked this out it does seem like body stream errors aren't reported how they should be... made a change to align the h1 reporting with how h2 does it: d9b31b80ac

**Reply by bburdette** [reply]
nice!  I'll look for that in a future release.
